### ISSUE TITLE
Subscription: renewal payment failed with Stripe Connect

### ISSUE THREAD

**Issue by ProRealCode** [weDevsOfficial/dokan#1261]
Hi, I have found that "sometimes", renewal payment failed through your Stripe Connect payment module, the error message is (under WC log & Stripe Log)
_caught exception on process_subscription_payment, code: unable_to_create_payment_intent, message: The provided source has already been used to make another payment. You cannot reuse sources without attaching them to a Customer object first._

Please find attached the picture from Stripe logs about the issue: 
<img width="1028" alt="stripe connect bug" src="https://user-images.githubusercontent.com/23722067/123632367-21093b80-d818-11eb-8887-31194cb70b29.png">

It happened 2 times already and with different account/customer.

Please provide assistance on how to resolve this issue, thanks.

**Comment by ProRealCode** [comment]
Some other renewal have successfully worked. Today another one has failed.

I have found that, for the orders that have failed, the "_stripe_source_id" card has a "consumed" status in Stripe and "Consumed sources cannot be used for further payments." (https://stripe.com/docs/sources/customers)

More over, I think the problem is as follow:

1/ client buy a 1st product with renewal, the source_id is created an attached to the customer

2/ client buy a 2nd product with renewal, he is using the same credit card and a new source_id is created an attached to the customer (seems to replace the current source_id)

3/ first renewal payment is due for the 1st product, Stripe doesnt recognize the credit card has the source_id is not the same anymore

It means that if a client has more than 1 product with a subscription, all subsequent payments will fail! This issue must be solved asap...

**Comment by nurul-umbhiya** [comment]
@shanjidah  can you please test the use cases mentioned in the above comment?

**Comment by nurul-umbhiya** [comment]
@ProRealCode We've taken this issue into considertaion. We will try to provide a solution as soon as possible.

**Comment by ProRealCode** [comment]
Thank you for feedback, so you were able to replicate the issue?

**Comment by ProRealCode** [comment]
Hi, I see that there is a new fix in the v3.3.4 version updated today: 
Card is Not Saving While Purchasing WooCommerce Subscription Products [Dokan Stripe]. Fix
[Dokan Stripe] Card is not saving while purchasing WooCommerce Subscription products (3ds/non3ds)

Why does it means for previous subscriptions made prio to this bug fix please? While I see some of the recurring payments are ok and some don't? How should I handle future failing recurring payment?

**Comment by nurul-umbhiya** [comment]
@ProRealCode I would suggest you to ask customers with no saved cards to add their card again from site.com/my-account/payment-methods, if they add cards, their payment will be successful, otherwise, failed orders will be on On-Hold state.

**Comment by ProRealCode** [comment]
thank you for the fast answer.

**Comment by nurul-umbhiya** [comment]
Since the problem has been fixed, I'm closing this issue.

**Comment by ProRealCode** [comment]
> @ProRealCode I would suggest you to ask customers with no saved cards to add their card again from site.com/my-account/payment-methods, if they add cards, their payment will be successful, otherwise, failed orders will be on On-Hold state.

Hi, regarding your tip to be sure clients have their credit cards saved, i'm struggle to find where is located in the DB (user_meta?) the information that a specific user has a saved card. Since I have many subscribers, that would ease the process of verification. Could you point me in the right direction: where is located that information please? Thank you again.

**Comment by nurul-umbhiya** [comment]
@ProRealCode Cards tokens are stored at `prefix_woocommerce_payment_tokens`  `prefix_woocommerce_payment_tokenmeta` tables.

**Comment by ProRealCode** [comment]
Thanks for the tip. I found many subscribers with no saved cards. Does it means that all of them will encounter a payment issue on the next renewal or only a few?

**Comment by ProRealCode** [comment]
Hi, the problem is still present today.
A customer which has subscribed before your recent update of the plugin encountered the same message error as before resulting in a non-payment of the subscription..
![image](https://user-images.githubusercontent.com/23722067/130442360-0082ebbb-9f0c-4670-8c38-003ba5e66a23.png)

This customer has a credit card attached as a default payment card in site.com/my-account/payment-methods , as you suggested.

I have numerous other subscription payments to come, please provide a solution to the issue quickly.

**Comment by nurul-umbhiya** [comment]
@ProRealCode customer needs to add card to their account, otherwise subscription won't work :(

**Comment by ProRealCode** [comment]
Sorry but the customer has a card saved in his account, but it was saved before your plugin update, is it the cause of the payment failure?

**Comment by ProRealCode** [comment]
Could you be more precise about the saved cards infos. Is it automatic for new customer when they buy a subscription product or not? If it is not, they do need to save their cards manually through their account dashboard? So far I have no clue on how to resolve the current issue for both new subscribers and old ones.

As a reminder, an old customer with a saved card can still have a payment failure like I stated in my previous message: https://github.com/weDevsOfficial/dokan/issues/1261#issuecomment-903694745

How are we supposed to run a business without knowing how is managed the payment process? Please..

**Comment by nurul-umbhiya** [comment]
> Is it automatic for new customer when they buy a subscription product or not?

Yes it is automated, If a customer purchase a subscription product, their purchased card information will be stored in Stripe end and in site woocommerce will store card token. next recurring billing will be taken from that stored card. Unfortunately, there was an issue where the customer card wasn't stored for a subscription payment. that is why payments are getting rejected from the stripe end. But we fixed it already. for users with no saved card, what I can suggest is: for old users without a saved cards, asked them to save their card from My Account --> Payment methods page and mark that card as default [.](https://prnt.sc/1qhv5de) Otherwise, recurring payments will fail.  

> So far I have no clue on how to resolve the current issue for both new subscribers and old ones.

With the latest version, you don't have to worry about card issue, it is working perfectly now

> As a reminder, an old customer with a saved card can still have a payment failure like I stated in my previous message: #1261 (comment)

This can happen for two reasons: 
1. Recurring subscription needs user consent to charge their cards for future payments, if the card wasn't stored for future use (in case of non-recurring products where admin settings for save card was enabled), this card can't be used to charge recurring payment.
2. If a user got multiple saved cards and no card is attached to a subscription, the default card will be charged, if the default card was added by a non-recurring payment (in this case, if the customer needs to use that card, they need to give their consent again), that card can't be used for recurring payments. In this case, the customer needs to make a card default which was added via My Account --> Payment Methods.

The bottom line is, for old customer, they need to add their card again via My Account --> Payment methods page and mark that card as default

**Comment by ProRealCode** [comment]
Thank you for your quick answer.

Everything is clear, except for the admin setting "Enable registered cards". Do you confirm that now, even if this setting is set to ON, if the customer purchases a non-recurring product, this setting will not affect their previously registered card to pay for their other previous subscription orders that will come due for payment? 

![image](https://user-images.githubusercontent.com/23722067/130796931-af65641a-b43e-4400-b1e7-5a4ba0a98a0d.png)

**Comment by nurul-umbhiya** [comment]
Subscription products will not get affected as card token is saved under parent subscription order.
